tsdb/exemplar_test.go

```diff
@@ -0,0 +1,538 @@
+}
+
+func TestSelectExemplarOrdering(t *testing.T) {
+	es, err := NewCircularExemplarStorage(5, nil)
+	require.NoError(t, err)
+
+	l := labels.Labels{
+		{Name: "service", Value: "asdf"},
+	}
+	exemplars := []exemplar.Exemplar{
+		{
+			Labels: labels.Labels{
+				labels.Label{
+					Name:  "traceID",
+					Value: "a",
+				},
+			},
+			Value: 0.1,
+			Ts:    101,
+		},
+		{
+			Labels: labels.Labels{
+				labels.Label{
+					Name:  "traceID",
+					Value: "b",
+				},
+			},
+			Value: 0.2,
+			Ts:    102,
+		},
+		{
+			Labels: labels.Labels{
+				labels.Label{
+					Name:  "traceID",
+					Value: "c",
+				},
+			},
+			Value: 0.3,
+			Ts:    103,
+		},
+		{
+			Labels: labels.Labels{
+				labels.Label{
+					Name:  "traceID",
+					Value: "d",
+				},
+			},
+			Value: 0.4,
+			Ts:    104,
+		},
+		{
+			Labels: labels.Labels{
+				labels.Label{
+					Name:  "traceID",
+					Value: "e",
+				},
+			},
+			Value: 0.5,
+			Ts:    105,
+		},
+		{
+			Labels: labels.Labels{
+				labels.Label{
+					Name:  "traceID",
+					Value: "f",
+				},
+			},
+			Value: 0.6,
+			Ts:    106,
+		},
+	}
+
+	for _, e := range exemplars {
+		es.AddExemplar(l, e)
+	}
+	require.True(t, reflect.DeepEqual(es.exemplars[0].exemplar, exemplars[5]), "exemplar was not stored correctly")
+
+	ret, err := es.Select(100, 110, l)
+	require.NoError(t, err)
+
+	require.True(t, reflect.DeepEqual(exemplars[1:], ret), "select did not return expected exemplars\n\texpected: %+v\n\tactual: %+v\n", exemplars, ret)
+}
+
+func TestSelectExemplar_Circ(t *testing.T) {
+	es, err := NewCircularExemplarStorage(3, nil)
+	require.NoError(t, err)
+
+	l := labels.Labels{
+		{Name: "service", Value: "asdf"},
+	}
+	exemplars := []exemplar.Exemplar{
+		{
+			Labels: labels.Labels{
+				labels.Label{
+					Name:  "traceID",
+					Value: "qwerty",
+				},
+			},
+			Value: 0.1,
+			Ts:    101,
+		},
+		{
+			Labels: labels.Labels{
+				labels.Label{
+					Name:  "traceID",
+					Value: "zxcvbn",
+				},
+			},
+			Value: 0.1,
+			Ts:    102,
+		},
+		{
+			Labels: labels.Labels{
+				labels.Label{
+					Name:  "traceID",
+					Value: "asdfgh",
+				},
+			},
+			Value: 0.1,
+			Ts:    103,
+		},
+	}
+
+	for i, e := range exemplars {
+		err := es.AddExemplar(l, e)
+		require.NoError(t, err)
+		require.Equal(t, es.index[l.String()].last, i, "exemplar was not stored correctly")
+	}
+
+	el, err := es.Select(100, 105, l)
+	require.NoError(t, err)
+	require.True(t, len(el) == 3, "didn't get expected one exemplar")
+
+	for i := range exemplars {
+		require.True(t, el[i].Equals(exemplars[i]), "returned exemplar did not matched expected, actual: %+v expected %+v", el[i], exemplars[i])
+	}
+}
+
+// This is a set of stored exemplars I scraped and stored locally that resulted in an infinite loop.
+// This test ensures Select doesn't infinitely loop on them anymore.
+func TestSelectExemplar_OverwriteLoop(t *testing.T) {
```

Comment:
1. It would be nice to use the `AddExemplar` to add exemplars instead of manually assigning - else it is not testing the real world like scenario. All exemplars have 0 `Ts`, so we are not testing a valid state of the queue.